File: src/de/sciencecomputing/venuside/BashMethodAnnotator.java
package de.sciencecomputing.venuside;

import de.sciencecomputing.venuside.psi.ContextAssignment;

import com.ansorgit.plugins.bash.lang.psi.api.BashPsiElement;
import com.ansorgit.plugins.bash.lang.psi.impl.word.BashWordImpl;
import com.ansorgit.plugins.bash.lang.psi.impl.command.BashSimpleCommandImpl;

import com.intellij.lang.annotation.Annotation;
import com.intellij.lang.annotation.AnnotationHolder;
import com.intellij.lang.annotation.Annotator;
import com.intellij.openapi.editor.DefaultLanguageHighlighterColors;
import com.intellij.codeInspection.ProblemHighlightType;
import com.intellij.psi.PsiElement;

import org.jetbrains.annotations.NotNull;

import java.util.List;

public class BashMethodAnnotator implements Annotator {
    @Override
    public void annotate(@NotNull final PsiElement element, @NotNull AnnotationHolder holder) {
        if (!(element instanceof BashSimpleCommandImpl))
            return;
        BashSimpleCommandImpl cmd = (BashSimpleCommandImpl) element;
        String cmdName = cmd.getReferencedCommandName();
        if (cmdName == null)
            return;
        String methodPath = DomainTree.depotPath(element.getContainingFile());
        if (methodPath == null) {
            holder.createWarningAnnotation(cmd.getTextRange(), "venus API call might be outside of method");
            return;
        }
        if (cmdName.equals("venus_import_variable"))
            annotateImportVariable(cmd, holder, methodPath);
    }

    private void annotateImportVariable(@NotNull BashSimpleCommandImpl cmd, @NotNull AnnotationHolder holder, @NotNull String methodPath) {
        BashWordImpl varWord = null;
        String contextName = cmd.getContainingFile().getName();
        for (BashPsiElement param: cmd.parameters()) {
            if (!(param instanceof BashWordImpl))
                return; // ???
            BashWordImpl paramWord = (BashWordImpl) param;
            if (!paramWord.isStatic())
                return; // could evaluate to any number of parameters
            if (contextName == null)
                contextName = param.getText();
            else if (param.getText().startsWith("-"))
                continue;
            else if (param.getText().equals("from")) {
                contextName = null;
                holder.createInfoAnnotation(param.getTextRange(), null)
                        .setTextAttributes(DefaultLanguageHighlighterColors.KEYWORD);
            } else
                varWord = paramWord;
        }
        if (varWord == null) {
            holder.createErrorAnnotation(cmd.getTextRange(), "variable import without variable name");
            return;
        } else if (contextName == null) {
            holder.createErrorAnnotation(cmd.getTextRange(), "venus_import_variable ... from without context name");
            return;
        }

        ContextAssignment var = DomainTree.lookupContextVariable(cmd.getProject(),
                methodPath, contextName, varWord.getText());
        if (var == null) {
            Annotation annotation = holder.createWeakWarningAnnotation(varWord.getTextRange(), "context variable might be undefined");
            annotation.setHighlightType(ProblemHighlightType.LIKE_UNKNOWN_SYMBOL);
        } else {
            Annotation annotation = holder.createInfoAnnotation(varWord.getTextRange(), var.getValue());
            annotation.setTextAttributes(DefaultLanguageHighlighterColors.GLOBAL_VARIABLE);
        }
    }
}

File: src/de/sciencecomputing/venuside/BashMethodAnnotator.kt
package de.sciencecomputing.venuside

import com.ansorgit.plugins.bash.lang.psi.impl.command.BashSimpleCommandImpl
import com.ansorgit.plugins.bash.lang.psi.impl.word.BashWordImpl

import com.intellij.codeInspection.ProblemHighlightType
import com.intellij.lang.annotation.AnnotationHolder
import com.intellij.lang.annotation.Annotator
import com.intellij.openapi.editor.DefaultLanguageHighlighterColors
import com.intellij.psi.PsiElement

class BashMethodAnnotator : Annotator {
    override fun annotate(element: PsiElement, holder: AnnotationHolder) {
        if (element !is BashSimpleCommandImpl)
            return
        val cmdName = element.referencedCommandName ?: return
        val methodPath = depotPath(element.containingFile)
        if (methodPath == null) {
            holder.createWarningAnnotation(element.textRange, "venus API call might be outside of method")
            return
        }
        if (cmdName == "venus_import_variable")
            annotateImportVariable(element, holder, methodPath)
    }

    private fun annotateImportVariable(cmd: BashSimpleCommandImpl, holder: AnnotationHolder, methodPath: String) {
        var varWord: BashWordImpl? = null
        var contextName: String? = cmd.containingFile.name
        for (param in cmd.parameters()) {
            if (param !is BashWordImpl)
                return  // ???
            if (!param.isStatic)
                return  // could evaluate to any number of parameters
            if (contextName == null)
                contextName = param.getText()
            else if (param.getText().startsWith("-"))
                continue
            else if (param.getText() == "from") {
                contextName = null
                holder.createInfoAnnotation(param.getTextRange(), null).textAttributes = DefaultLanguageHighlighterColors.KEYWORD
            } else
                varWord = param
        }
        if (varWord == null) {
            holder.createErrorAnnotation(cmd.textRange, "variable import without variable name")
            return
        } else if (contextName == null) {
            holder.createErrorAnnotation(cmd.textRange, "venus_import_variable ... from without context name")
            return
        }

        val variable = lookupContextVariable(cmd.project, methodPath, contextName, varWord.text)
        if (variable == null)
            holder.createWeakWarningAnnotation(varWord.textRange, "context variable might be undefined")
                .highlightType = ProblemHighlightType.LIKE_UNKNOWN_SYMBOL
        else
            holder.createInfoAnnotation(varWord.textRange, variable.value)
                .textAttributes = DefaultLanguageHighlighterColors.GLOBAL_VARIABLE
    }
}
